Vamp docs and kmeans performance improvements

clonker opened the pull request:
- draft for vamp docs (very wip)
- kmeans++ re-implementation

brookehus commented on sktime/clustering/kmeans.py:
```diff
@@ -37,13 +40,13 @@ def __init__(self, n_clusters, cluster_centers, metric, tolerance, inertia=np.in
             Tolerance which was used as convergence criterium.
-        inertia : float, optional, default=np.inf
-            Value of the cost function at the end of the estimation.
+        inertias : (t,) ndarray, optional, default=None
```
I don't really understand from the docs what happens if I put in an array instead of `None`

clonker replied:
added clarifying docstring, this is basically just data that falls out of the estimator. The aim is to have model classes which can be instantiated without estimator as well - in this case based on cluster centers and metric.

brookehus replied:
cool!

brookehus commented on sktime/data/datasets.py:
```diff
@@ -45,3 +48,50 @@ def ellipsoids(laziness=0.97, seed=None):
     return Ellipsoids(laziness=laziness, seed=seed)
+
+
+def pbf(n_burn_in=5000, n_jobs=None):
```
would love a clearer name

clonker replied:
called `position_based_fluids` now

brookehus replied:
nice!!

brookehus commented on sktime/data/datasets.py:
```diff
@@ -45,3 +48,50 @@ def ellipsoids(laziness=0.97, seed=None):
+    init_pos_y = np.arange(-12, 24, interaction_distance * .9).astype(np.float32)
+    init_pos = np.dstack(np.meshgrid(init_pos_x, init_pos_y)).reshape(-1, 2)
+    domain = np.array([80, 50])
+    pbf = PBF(domain_size=domain, initial_positions=init_pos, interaction_distance=interaction_distance, n_jobs=n_jobs)
```
variable `pbf` as an instance of class `PBF` within method `pbf` is not my favorite :)

clonker replied:
Renamed `PBF` to `PBFSimulator`, so `pbf = PBFSimulator(...)` is hopefully acceptable. :)

brookehus replied:
yesss

brookehus commented on sktime/data/datasets.py:
```diff
@@ -45,3 +48,50 @@ def ellipsoids(laziness=0.97, seed=None):
+
+    The simulation box has dimensions :math:`[-40, 40]\times [-25, 25]` and the initial positions are positioned
+    around the top boundary of the box. The interaction distance is set to :math:`d = 1.5` and `n_burn_in` steps are
+    performed to equilibrate the system before returning the simulator.
```
I would love more details about how the # of particles (which I think are # of initial positions?) is determined

clonker replied:
the `datasets.py` file provides a top-level interface to the simulator. Here, the simulation domain as well as number of particles (in terms of initial positions) are fixed. In case custom positioning is wanted, one should refer to the simulator. I have clarified this in the docstring now.

brookehus replied:
yay

brookehus commented on sktime/data/pbf.py:
```diff
@@ -0,0 +1,312 @@
+        domain_size : (2,) ndarray
+            A 1-dimensional ndarray with two elements describing the extension of the simulation box.
+        initial_positions : (N, 2) ndarray
+            A 2-dimensional ndarray describing the :math:`N` particles' initial positions.
```
this is exactly how the user specifies number of particles, right? by the shape of this array?

clonker replied:
yes, added clarifying docstring

brookehus replied:
thank you!

brookehus commented on sktime/data/pbf.py:
```diff
@@ -0,0 +1,312 @@
+            The number of solver iterations for particle position updates. The more, the slower the simulation but also
+            the more accurate.
+        gravity : float, default=10
+            Gravity parameter which acts on particles' velolicites in each time step as constant force.
```
typo velocities

brookehus commented on sktime/data/pbf.py:
```diff
@@ -0,0 +1,312 @@
+        tensile_instability_k : float, default=0.1
+            Also controls surface-tension effects.
+        """
+        if np.atleast_1d(domain_size).ndim != 1 or domain_size.shape[0] != 2 or np.any(domain_size <= 0):
```
great numpy function

brookehus commented on sktime/decomposition/vamp.py:
```diff
@@ -60,7 +59,7 @@ def __init__(self, cov: CovarianceModel, dim=None, epsilon=1e-6, scaling=None, r
         scaling : str or None, optional, default=None
             Scaling parameter, see :attr:`VAMP.scaling`.
-        right : bool, optional, default=True
+        right : bool, optional, default=False
```
so does changing this default to false mean the left eigenvectors are used for projection? or...?

clonker replied:
This is usually set from the estimator, it is just so that the defaults from the estimator match the defaults from the model. defaulting `right=False` means that not the right but the left eigenvectors are used, indeed

brookehus replied:
might be nice to write explicitly since making it true/false w.r.t. one set is a bit hard to grok

clonker replied:
I have added a docstring. Normally I would say that this should not be a parameter at all but rather be specified whenever transforming data. However to obey the sklearn interface, it must be a state variable.

brookehus replied:
thank you!

brookehus commented on sktime/decomposition/vamp.py:
```diff
@@ -347,6 +338,42 @@ def _decomposition(self) -> _DiagonalizationResults:
+            The input trajectory
+        component : int or None
+            The component to project onto. If None, all processes are taken into account, if integer sets all singular
+            values to zero but the "component"th one.
```
is there ever an interest in using more than one but less than all

clonker replied:
perhaps for superposition of components, ill add it to the implementation

brookehus replied:
seems straightforward to add, right?

clonker replied:
quite straightforward, added and also modified the vamp notebook a bit to showcase a usage of multiple components

brookehus replied:
this is great!

brookehus commented on sktime/markov/msm/markov_state_model.py:
```diff
@@ -68,10 +68,14 @@ def __init__(self, transition_matrix: np.ndarray,
         count_model : TransitionCountModel, optional, default=None
             In case the MSM was estimated from data, the transition count model can be provided for statistical
             information about the data. Some properties of the model require a count model so that they can be computed.
+        transition_matrix_tolerance : float, default=1e-8
```
nice!

brookehus commented on sktime/markov/msm/markov_state_model.py:
```diff
@@ -133,16 +137,23 @@ def transition_matrix(self):
-            if not msmana.is_transition_matrix(value, tol=1e-8):
-                raise ValueError('The input transition matrix is not a stochastic matrix '
-                                 '(elements >= 0, rows sum up to 1).')
+            if not msmana.is_transition_matrix(value, tol=self.transition_matrix_tolerance):
```
whats `msmana`

clonker replied:
an `msmtools` import, bound to change anyways

brookehus replied:
cool

brookehus commented on sktime/util.py:
```diff
@@ -1,15 +1,38 @@
+    In particular, if
+
+      * value is None, use 1 job
+      * value is negative, use number cores available * 2
```
0_o

clonker replied:
This yields better load-balancing and makes better use of hyperthreading

brookehus replied:
cool

brookehus commented:
Notebook notes:

[TICA vs VAMP](https://mhoffmann.xyz/sktime/index_dimreduction.html):
- I would note somewhere in the last paragraph of "When to use TICA" that TICA can also be a problem for simulations with rare events: i.e. the underlying distribution is expected to obey microscopic reversibility, but not enough sampling has been performed (I see that you mention this below in the VAMP section but I think repeating here would be good)

Did not re-look @ clustering nb


[VAMP notebook](https://mhoffmann.xyz/sktime/notebooks/vamp.html#Example-with-position-based-fluids)
- `pbf = sktime.data.pbf(n_jobs=8)`: Why is the sktime bit not called `PBFSimulator` or something? If you don't feel like that, at least the variable name on the LHS can be more clear
- Why is `PBF` capitalized (here)[https://mhoffmann.xyz/sktime/api/generated/impl/sktime.data.PBF.html#sktime.data.PBF] but not in the notebook?
- How do we get/know the number of particles? In the PBF docs it doesn't seem specifiable? Is it the size of the initial positions arg? Then why are initial positions not input in the VAMP notebook code snippet?
- I love the imported video!!!!
- Can the PBF be seeded?
- If easy, bigger titles/axis labels/ticks on the plot after box 25?
- `the second slowest process is associated wtih a density change middle <-> sides` (typo @ with)
- Propagating into the future is so cool!!!!

Did not look at `pbf.h`

Did not look at `sktime/src/include/common.h` or `thread_utils.h`

clonker commented:
Thanks for the review! Addressing some of the points:
- PBF is now called PBFSimulator, the dataset api method is called `position_based_fluids`.
- In the dataset the positions are predetermined and not placed randomly, the simulator itself is (up to machine precision) deterministic. So there is no seeding possible / required.
- The number of particles is determined by the number of initial positions, i.e. the shape of the position array. This number is preserved during a simulation.

brookehus commented on sktime/clustering/kmeans.py:
```diff
@@ -38,10 +39,12 @@ def __init__(self, n_clusters, cluster_centers, metric, tolerance, inertias=None
+            can be constructed purely from cluster centers and metric.
+        inertias : (t,) ndarray or None, optional, default=None
+            Value of the cost function over :code:`t` iterations. Defaults to `None` so that clustering models
+            can be constructed purely from cluster centers and metic.
```
metric?

clonker replied:
metric!

brookehus commented:
If there is not a note about the same pbf params giving the same sim, to machine precision, that might be nice to add. Otherwise LGTM and your changes are great!
